### ISSUE TITLE
EnvFrom semantics might be unexpected

### ISSUE THREAD

**Issue by pmorie** [kubernetes/kubernetes#42583]
TLDR: I have had this nagging feeling that the semantics of `EnvFrom` are wrong.  I think we should either change them to be what users would expect, *or* ensure that the semantics are fully represented in all user-facing doc.

**Is this a request for help?**

No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.):

EnvFrom

---

**Kubernetes version** (use `kubectl version`):

1.6 master

**Environment**:

All environments

**What happened**:

When using `EnvFrom` with a Secret or ConfigMap, if the resource contains any non-environment keys (dashes, etc), then my pod does not start.

**What you expected to happen**:

I expect to only get keys that are legal environment variables from a resource I use with this feature and for it not to affect whether my pod starts.  You should get an event on the pod saying that X or Y keys were omitted because they are not valid environment variables.

**How to reproduce it** (as minimally and precisely as possible):

Make a secret or configmap with 2 environment-variable-compatible keys and 1 non-compatible key.  Depend on it in a pod using `EnvFrom`.  The pod won't start.

**Anything else we need to know**:

This behavior re: keys that are not valid environment variable names is documented in the `types.go` files but not the Kubernetes web documentation: https://github.com/kubernetes/kubernetes.github.io/pull/1967

**Comment by derekwaynecarr** [comment]
I think we should omit invalid keys and possibly throw an event with an aggregate list of keys that were not projected.

**Comment by fraenkel** [comment]
This is a bit of a slippery slope. We don't talk about valid or invalid anything when it comes to config maps or environment variables in the documentation.

Right now we fail if the ConfigMap isn't present or if the key within a ConfigMap isn't there.

What is the difference between an invalid key, a missing key or missing ConfigMap?

**Comment by pmorie** [comment]
> We don't talk about valid or invalid anything when it comes to config maps or environment variables in the documentation.

If that is the case, we should fix it.

> Right now we fail if the ConfigMap isn't present or if the key within a ConfigMap isn't there.

That is accurate, let me pick apart my thoughts about these behaviors:

1.  It is right to fail if the named resource (secret/configmap) doesn't exist, **unless** the reference to it in the podspec is marked as optional, because you explicitly said that you depend on the resource
2.  It is also right to fail in `env` if a key that you **explicitly named**, which is not optional, does not exist

In the case of `EnvFrom`, here's how I see things:

1.  If you use `EnvFrom` to consume a resource, and that resource is not marked optional, and that resource does not exist, we should fail
2.  If you use `EnvFrom` to consume a resource, but one of that resource's keys is not in the right format, you should start the pod and create an event listing the keys from that resource that were skipped

The rationale for (2) is that you did not explicitly name a key.  Therefore, you didn't make a statement that you needed a particular key.  I think also that name `EnvFrom` makes it sound like you will be injected only with the keys that are environment variables from the resource.

Thoughts?

**Comment by fraenkel** [comment]
That would work. It is easy enough to explain. Will have to look at the docs, to see where we should do this.
I assume it also applies to Secrets.

Is this for 1.6? or future?

**Comment by pmorie** [comment]
@fraenkel Cool.  I would like to get this resolved in 1.6 so that we have a good, consistent story from the moment the feature is released.

**Comment by pmorie** [comment]
@fraenkel We should make sure to update docs with examples showing what happens with keys that don't adhere to the format.
